designs/2020-es-module-support/README.md

```diff
@@ -0,0 +1,159 @@
+* every module has it's own unique build system, with no consistency across modules.
+* projects are packed with complex configurations, much of which is plumbing related to packaging/bundling.
+
+Over time, browsers have adopted the ecma module standard. Today, support is pretty good:
+https://caniuse.com/es6-module
+
+Node also added support in v12.17.1 and up, and has officially marked the API stable:
+https://nodejs.org/dist/latest-v15.x/docs/api/esm.html#esm_modules_ecmascript_modules
+
+
+By adopting this new standard, we can achieve a few things:
+
+* eventual simplification of espree's build process by eliminating the need for multiple module formats
+* make it easier to directly importing espree in browser environments, deno, etc. Skypack is a CDN that already enables this, but they have internal logic that looks at every package in npm and tries to build an es module. https://www.skypack.dev/
+
+As more projects adopt this standardized module format, the hope is a lot of this "connective glue" will fall away, leaving us with leaner modules and less build tooling across our stack.
+
+
+## Detailed Design
+
+### Nomenclature
+
+* `ESM` ecmascript module. The standardized module format for javascript
+* `CJS` commonjs. Node's historical module format (i.e., `module.exports = ...`, `const a = require('foo')`)
+* `UMD` universal module definition. A legacy format for specifying javascript code as a module in a way that tries to guarantee compatibility with many loader formats including requireJS, commonjs, and global script tags.
+* `IIFE` immediately invoked function expressions. A legacy strategy for encapsulating javascript into a module that won't leak global variables into the execution environment. It is still the primary way of packaging global script tags for inclusion in a web page. `UMD` is usually wrapped in an `IIFE`
+* `CDN` content delivery network. can be thought of as a set of servers that host static content all around the globe to deliver assets quickly and at scale.
+
+
+### build process
+
+Today, espree is written in commonjs (CJS) format, and a universal module definition (UMD) build is created too.
+
+
+today's build process:
+```
+┌-------------------┐               ┌-----------------┐
+│                   │               │                 │                package.json:
+│     espree.js     |               │ build/espree.js │
+│ (CJS entry point) ├──BUILD_STEP──▶│   (UMD bundle)  │     CJS ---▶   "main": "espress.js",
+│                   │               │                 │
+└-------------------┘               └-----------------┘
+```
+
+
+The strategy that probably makes the most sense in this initial work is to do what is known as a "dual package" solution, where
+espree provides an ESM, and a CJS module.
+
+Providing a dual package shouldn't affect any existing espree users, but provies an ESM option for users that want it. 
+
+
+proposed build process in this RFC:
```

Comment:
I'd like to suggest a change here. Instead of bundling a CJS version and just shipping the ESM source code, I'd like to see us generate:

* `dist/espree.cjs` - the CJS version
* `dist/espree.js` - the ESM version

That way, we'd be shipping the exact equivalent between CJS and ESM, and we can just include `dist` in the `files` section of `package.json`.

Reply:
what would go into `dist/espree.js`, is this an identical file to what is found in `espree.js` ? My thinking is that the build step is just a temporary condition, and eventually `dist/` becomes unnecessary.

Reply:
@nzakas would these be the thin wrappers described in https://nodejs.org/api/packages.html#packages_approach_1_use_an_es_module_wrapper or full compiled builds? The wrapper approach could avoid a build step entirely, which would be nice.

Reply:
Looks like we've agreed to move forward with the build step for now, so marking as resolved.